## <center>**Assignment**
# <center>**Hyperparameter Tuning**
Given credit_card_default dataset<br>
• Do the pre-processing on dataset<br>
• Perform different algorithms such as kNN,SVM, Gradient Boosting, Ada
boosting and XG Boost<br>
• Also perform hyperparameter tuning<br>

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
data = pd.read_csv('/content/drive/MyDrive/datasets/credit_card_default.csv')

In [5]:
data.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  PAY_4  \
0   1      20000    2          2         1   24      2      2     -1     -1   
1   2     120000    2          2         2   26     -1      2      0      0   
2   3      90000    2          2         2   34      0      0      0      0   
3   4      50000    2          2         1   37      0      0      0      0   
4   5      50000    1          2         1   57     -1      0     -1      0   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...          0          0          0         0       689         0   
1  ...       3272       3455       3261         0      1000      1000   
2  ...      14331      14948      15549      1518      1500      1000   
3  ...      28314      28959      29547      2000      2019      1200   
4  ...      20940      19146      19131      2000     36681     10000   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0                           1  
1      1000         0      2000                           1  
2      1000      1000      5000                           0  
3      1100      1069      1000                           0  
4      9000       689       679                           0  

[5 rows x 25 columns]

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   ID                          30000 non-null  int64
 1   LIMIT_BAL                   30000 non-null  int64
 2   SEX                         30000 non-null  int64
 3   EDUCATION                   30000 non-null  int64
 4   MARRIAGE                    30000 non-null  int64
 5   AGE                         30000 non-null  int64
 6   PAY_1                       30000 non-null  int64
 7   PAY_2                       30000 non-null  int64
 8   PAY_3                       30000 non-null  int64
 9   PAY_4                       30000 non-null  int64
 10  PAY_5                       30000 non-null  int64
 11  PAY_6                       30000 non-null  int64
 12  BILL_AMT1                   30000 non-null  int64
 13  BILL_AMT2                   30000 non-null  int64
 14  BILL_A

In [7]:
data.describe()

ID       LIMIT_BAL           SEX     EDUCATION      MARRIAGE  \
count  30000.000000    30000.000000  30000.000000  30000.000000  30000.000000   
mean   15000.500000   167484.322667      1.603733      1.853133      1.553667   
std     8660.398374   129747.661567      0.489129      0.790349      0.518326   
min        1.000000    10000.000000      1.000000      0.000000      1.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22500.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                AGE         PAY_1         PAY_2         PAY_3         PAY_4  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean      35.485500     -0.016700     -0.133767     -0.166200     -0.220667   
std        9.217904      1.123802      1.197186      1.196868      1.169139   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...      BILL_AMT4      BILL_AMT5      BILL_AMT6       PAY_AMT1  \
count  ...   30000.000000   30000.000000   30000.000000   30000.000000   
mean   ...   43262.948967   40311.400967   38871.760400    5663.580500   
std    ...   64332.856134   60797.155770   59554.107537   16563.280354   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2326.750000    1763.000000    1256.000000    1000.000000   
50%    ...   19052.000000   18104.500000   17071.000000    2100.000000   
75%    ...   54506.000000   50190.500000   49198.250000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

           PAY_AMT2      PAY_AMT3       PAY_AMT4       PAY_AMT5  \
count  3.000000e+04   30000.00000   30000.000000   30000.000000   
mean   5.921163e+03    5225.68150    4826.076867    4799.387633   
std    2.304087e+04   17606.96147   15666.159744   15278.305679   
min    0.000000e+00       0.00000       0.000000       0.000000   
25%    8.330000e+02     390.00000     296.000000     252.500000   
50%    2.009000e+03    1800.00000    1500.000000    1500.000000   
75%    5.000000e+03    4505.00000    4013.250000    4031.500000   
max    1.684259e+06  896040.00000  621000.000000  426529.000000   

            PAY_AMT6  default payment next month  
count   30000.000000                30000.000000  
mean     5215.502567                    0.221200  
std     17777.465775                    0.415062  
min         0.000000                    0.000000  
25%       117.750000                    0.000000  
50%      1500.000000                    0.000000  
75%      4000.000000                    0.000000  
max    528666.000000                    1.000000  

[8 rows x 25 columns]

#### Spliting the data into training and testing sets

In [8]:
X = data.drop(columns=['ID', 'default payment next month'])
y = data['default payment next month']

In [9]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#### Standardizing the features using standardscaler

In [10]:
from sklearn.preprocessing import StandardScaler
scale = StandardScaler()
x_train_scaled = scale.fit_transform(x_train)
x_test_scaled = scale.fit_transform(x_test)

### Model Creation

#### KNN

In [11]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
knn.fit(x_train_scaled, y_train)
y_pred_knn = knn.predict(x_test_scaled)

#### SVM

In [12]:
from sklearn.svm import SVC
svm = SVC()
svm.fit(x_train_scaled,y_train)
y_pred_svm = svm.predict(x_test_scaled)

#### Gradient Boosting

In [13]:
from sklearn.ensemble import GradientBoostingClassifier, AdaBoostClassifier
grd_clf = GradientBoostingClassifier()
grd_clf.fit(x_train_scaled, y_train)
y_pred_grd_clf = grd_clf.predict(x_test_scaled)

#### Ada Boosting

In [14]:
ada_clf = AdaBoostClassifier()
ada_clf.fit(x_train_scaled, y_train)
y_pred_ada_clf = ada_clf.predict(x_test_scaled)

#### XG Boost

In [15]:
from xgboost import XGBClassifier
xgb_clf = XGBClassifier()
xgb_clf.fit(x_train_scaled, y_train)
y_pred_xgb_clf = xgb_clf.predict(x_test_scaled)

#### Metrics of different models

In [16]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

dt_ = [{'Accuracy': accuracy_score(y_test, y_pred_knn), 'Precision': precision_score(y_test, y_pred_knn),
       'Recall': recall_score(y_test, y_pred_knn), 'F1 Score': f1_score(y_test, y_pred_knn),
       'Confusion Matrix':confusion_matrix(y_test, y_pred_knn)},
      {'Accuracy': accuracy_score(y_test, y_pred_svm), 'Precision': precision_score(y_test, y_pred_svm),
       'Recall': recall_score(y_test, y_pred_svm), 'F1 Score': f1_score(y_test, y_pred_svm),
       'Confusion Matrix':confusion_matrix(y_test, y_pred_svm)},
      {'Accuracy': accuracy_score(y_test, y_pred_grd_clf), 'Precision': precision_score(y_test, y_pred_grd_clf),
       'Recall': recall_score(y_test, y_pred_grd_clf), 'F1 Score': f1_score(y_test, y_pred_grd_clf),
       'Confusion Matrix':confusion_matrix(y_test, y_pred_grd_clf)},
       {'Accuracy': accuracy_score(y_test, y_pred_ada_clf), 'Precision': precision_score(y_test, y_pred_ada_clf),
       'Recall': recall_score(y_test, y_pred_ada_clf), 'F1 Score': f1_score(y_test, y_pred_ada_clf),
       'Confusion Matrix':confusion_matrix(y_test, y_pred_ada_clf)},
       {'Accuracy': accuracy_score(y_test, y_pred_xgb_clf), 'Precision': precision_score(y_test, y_pred_xgb_clf),
       'Recall': recall_score(y_test, y_pred_xgb_clf), 'F1 Score': f1_score(y_test, y_pred_xgb_clf),
       'Confusion Matrix':confusion_matrix(y_test, y_pred_xgb_clf)},
     ]

index_labels=['KNN','SVM','Gadient Boost', 'Ada Boost', 'XGBoost']
df = pd.DataFrame(dt_, index=index_labels)

df

Accuracy  Precision    Recall  F1 Score  \
KNN            0.794167   0.545244  0.357959  0.432184   
SVM            0.818833   0.675466  0.331302  0.444558   
Gadient Boost  0.819333   0.659249  0.361005  0.466535   
Ada Boost      0.816667   0.666147  0.325209  0.437052   
XGBoost        0.806500   0.598958  0.350343  0.442095   

                        Confusion Matrix  
KNN            [[4295, 392], [843, 470]]  
SVM            [[4478, 209], [878, 435]]  
Gadient Boost  [[4442, 245], [839, 474]]  
Ada Boost      [[4473, 214], [886, 427]]  
XGBoost        [[4379, 308], [853, 460]]

### Hyperparameter Tuning

In [17]:
from sklearn.model_selection import GridSearchCV

#### KNN

We will use three hyperparamters- n-neighbors, weights and metric.<br>
*n_neighbors*: Decide the best k based on the values we have computed earlier.<br>
*weights*: Check whether adding weights to the data points is beneficial to the model or not. 'uniform' assigns no weight, while 'distance' weighs points by the inverse of their distances meaning nearer points will have more weight than the farther points.<br>
*metric*: The distance metric to be used will calculating the similarity.<br>

In [18]:
params_knn = {'n_neighbors': [3, 5, 7, 9,11], 'weights': ['uniform', 'distance'], 'metric' : ['minkowski','euclidean','manhattan']}
# Create GridSearchCV
grid_search_knn = GridSearchCV(knn, params_knn, cv=5, scoring='accuracy')
grid_search_knn.fit(x_train_scaled, y_train)
best_params_knn = grid_search_knn.best_params_
print("Best Parameters for kNN:", best_params_knn)
y_pred_knn_tuned = grid_search_knn.predict(x_test_scaled)

Best Parameters for kNN: {'metric': 'minkowski', 'n_neighbors': 11, 'weights': 'uniform'}


#### SVM

When training an SVM with the Radial Basis Function (RBF) kernel, two parameters must be considered: *C* and *gamma*. The parameter *C*, common to all SVM kernels, trades off misclassification of training examples against simplicity of the decision surface. A low *C* makes the decision surface smooth, while a high C aims at classifying all training examples correctly. *gamma* defines how much influence a single training example has. The larger *gamma* is, the closer other examples must be to be affected.

In [24]:
params_svm = {'C': [0.1, 1, 10], 'gamma': [0.1, 0.01, 0.001], 'kernel': ['rbf', 'linear']}
grid_search_svm = GridSearchCV(svm, params_svm, cv=5, scoring='accuracy')
grid_search_svm.fit(x_train_scaled, y_train)
best_params_svm = grid_search_svm.best_params_
print("Best Parameters for SVM:", best_params_svm)
y_pred_svm_tuned = grid_search_svm.predict(x_test_scaled)

KeyboardInterrupt: ignored

**NB:** <i>Taking hours to complete tuning of SVM so skiping this here.

#### Gradient Boosting<br><br>
learning_rate shrinks the contribution of each tree by learning_rate.<br>
n_estimators represents the number of trees in the forest. Usually the higher the number of trees the better to learn the data. However, adding a lot of trees can slow down the training process considerably, therefore we do a parameter search to find the sweet spot.

In [20]:
params_grb = {'n_estimators': [50, 100, 150], 'learning_rate': [0.05, 0.1, 0.2]}
grid_search_grb = GridSearchCV(grd_clf, params_grb, cv=5, scoring='accuracy')
grid_search_grb.fit(x_train_scaled, y_train)
best_params_grb = grid_search_grb.best_params_
print("Best Parameters for Gradient Boosting:", best_params_grb)
y_pred_grb_tuned = grid_search_grb.predict(x_test_scaled)

Best Parameters for Gradient Boosting: {'learning_rate': 0.05, 'n_estimators': 150}


#### AdaBoost

In [21]:
params_ada = {'n_estimators': [50, 100, 150], 'learning_rate': [0.05, 0.1, 0.2]}
grid_search_ada = GridSearchCV(ada_clf, params_ada, cv=5, scoring='accuracy')
grid_search_ada.fit(x_train_scaled, y_train)
best_params_ada = grid_search_ada.best_params_
print("Best Parameters for Ada Boosting:", best_params_ada)
y_pred_ada_tuned = grid_search_ada.predict(x_test_scaled)

Best Parameters for Ada Boosting: {'learning_rate': 0.1, 'n_estimators': 150}


#### XGBoost

In [22]:
params_xgb = {'n_estimators': [50, 100, 150], 'learning_rate': [0.05, 0.1, 0.2], 'max_depth': [3, 4, 5]}
grid_search_xgb = GridSearchCV(xgb_clf, params_xgb, cv=5, scoring='accuracy')
grid_search_xgb.fit(x_train_scaled, y_train)
best_params_xgb = grid_search_xgb.best_params_
print("Best Parameters for XGBoost:", best_params_xgb)
y_pred_xgb_tuned = grid_search_xgb.predict(x_test_scaled)

Best Parameters for XGBoost: {'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 50}


In [27]:
dt_ = [{'Best Parameters': best_params_knn, 'Accuracy': accuracy_score(y_test, y_pred_knn_tuned), 'Best Score': grid_search_knn.best_score_  },
      {'Best Parameters': 'NA', 'Accuracy':  'NA', 'Best Score': 'NA' },                        # accuracy_score(y_test, y_pred_svm_tuned), 'Best Score': grid_search_svm.best_score_ },
      {'Best Parameters': best_params_grb, 'Accuracy': accuracy_score(y_test, y_pred_grb_tuned), 'Best Score': grid_search_grb.best_score_  },
      {'Best Parameters': best_params_ada, 'Accuracy': accuracy_score(y_test, y_pred_ada_tuned), 'Best Score': grid_search_ada.best_score_  },
      {'Best Parameters': best_params_xgb, 'Accuracy': accuracy_score(y_test, y_pred_xgb_tuned), 'Best Score': grid_search_xgb.best_score_  },
     ]

index_labels=['KNN','SVM','Gadient Boost', 'Ada Boost', 'XGBoost']
pd.set_option('display.max_colwidth', None)
df = pd.DataFrame(dt_, index=index_labels)

df

Best Parameters  \
KNN            {'metric': 'minkowski', 'n_neighbors': 11, 'weights': 'uniform'}   
SVM                                                                          NA   
Gadient Boost                      {'learning_rate': 0.05, 'n_estimators': 150}   
Ada Boost                           {'learning_rate': 0.1, 'n_estimators': 150}   
XGBoost             {'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 50}   

               Accuracy Best Score  
KNN            0.811833   0.806458  
SVM                  NA         NA  
Gadient Boost  0.818667      0.822  
Ada Boost         0.818   0.817833  
XGBoost        0.820833      0.822

***Now we have improved the accuracy of al models especially for knn and XGBoost after hyperparameter tuning. <br>
The highest accuracy of 82% is got fo XGBoost.***